rclcpp/src/rclcpp/executors/detail/timer_manager.hpp

```diff
@@ -0,0 +1,356 @@
+#pragma once
+#include <chrono>
+#include <functional>
+#include <memory>
+#include <vector>
+#include <map>
+
+#include <rcl/timer.h>
+#include <rclcpp/timer.hpp>
+
+#include <inttypes.h>
+
+namespace rclcpp::executors
+{
+
+/**
+ * @brief A class for managing a queue of timers
+ *
+ * This class holds a queue of timers of one type (RCL_ROS_TIME, RCL_SYSTEM_TIME or RCL_STEADY_TIME).
+ * The queue itself manages an internal map of the timers, orders by the next time a timer will be
+ * ready. Each time a timer is ready, a callback will be called from the internal thread.
+ */
+class TimerQueue
+{
+  struct TimerData
+  {
+    std::shared_ptr<const rcl_timer_t> rcl_ref;
+    std::function<void()> timer_ready_callback;
+  };
+
+public:
+  TimerQueue(rcl_clock_type_t timer_type)
+  : timer_type(timer_type),
```

Comment:
rather than storing this as an additional member, we could just use the `get_clock_type()` method of the clock